rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+Managed environment in that *As a Service* not only basic health but even
+optimization and proactive checking for problems with the service level.  A
+challenge here is that some of the information about service health is not
+immediately available from within the Capella Data Plane; it is in the user's
+application deployment in the cloud.  That said, some users have not defined
+service levels and are not measuring them, but rather just respond when they
+receive complaints or perceive that service levels are outside normal operating
+parameters.
+
+# General Design
+
+This feature likely will not, in many cases, point directly to the cause of an
+issue.  In most cases, it will help to ensure availability of the metrics,
+particularly when other the application does to retain logs for a long enough
+period.
+
+To provide the most value at the lowest cost, two levels of metrics are
+proposed:
+
+* Capella Service Experience Counters
+* Capella Service Level Metrics
+
+## Capella Service Experience Counters
+
+As initial background to those who are not familiar with the [SDK-RFC on error
+handling](rfc/0058-error-handling.md), Couchbase SDKs attempt to consolidate
+much of the error handling for an app developer into a few specific errors many
+of which may have an error context. The SDKs try to balance this Couchbase
+design of error handling with respecting platform idioms.
+
+That means in SDK instances, we largely have three major errors that a user may
+see along with additional error context.  The error context tends to be service
+specific and these may not have a similar shape as the services were
+independently developed (among other reasons).  There are a number of other
+errors, like InvalidArgument and KeyNotFound which are not useful from a
+Capella Experience perspective.
+
+The three metrics we should collect are:
+
+* Ambiguous Timeout
+  * The requested operation has timed out, but the SDK does not know if it was
+  received by the intended remote system.  For example, it was written to the
+  network and timed out before the response was received.  This is only for
+  mutations, as get() operations will always return an UnambiguiousTimeout.
+* Unambiguous Timeout
+  * The requested operation has timed out and the SDK knows it could not have
+  been received by the intended remote system.  For example, it was intended
+  for a system which there was no connection for and the connection couldn't be
+  built to send it before the connection timeout.
+* Request Canceled
+  * The requested operation was started and dispatched to the intended system.
+  However, it has been canceled by the SDK's processing.  Typically this is if
+  the connection is severed, but it could also have other causes (e.g.,
+  shutdown) and can even be triggered by the user themselves (e.g., calling
+  `.cancel()` on a result from another thread)
+
+An additional metric for total requests completed would allow to account for
+other types of errors.
```

Comment:
Would the total requests completed counter allow us to account for other error types? We don't distinguish between successful & unsuccessful requests.

Reply:
No, at this stage we do not want to distinguish any other errors except timeouts and cancellations.

Reply:
This was mostly a comment in the wording here. The total requests counter doesn't tell us anything about how many errors occurred which is what the lines here indicate:
> An additional metric for total requests completed would allow to account for
other types of errors.